Add spot instances docs for Azure (#796)
giantswarm/docs: src/content/advanced/spot-instances/azure/ondemand-fallback/index.md

---
linkTitle: Using on-demand as fall-back on Azure
title: Using on-demand instances as fall-back when spot instances are unavailable on Azure
description: When using spot instances in a node pool on Azure, it can happen that the node pool cannot be scaled up as not enough spot instances are available. This guide shows you how to configure cluster-autoscaler in a way to provide on-demand instances as a back-up automatically.
weight: 30
menu:
  main:
    parent: advanced-spotinstances-azure
user_questions:
  - How can I use VM spot instances and fall back to on-demand if no spot instances are available?
aliases:
  - /advanced/spot-instances/azure/on-demand-fallback/
owner:
  - https://github.com/orgs/giantswarm/teams/team-celestial
---

# Using on-demand instances as fall-back when spot instances are unavailable

> As of workload cluster release v{{% first_azure_spotinstances_version %}} for Azure, node pools can use [spot instances]({{< relref "/advanced/spot-instances/azure/overview" >}}). Users can select for each node pool which percentage of nodes should be covered by spot instances and by on-demand instances respectively. And users can define a base amount of on-demand instances that will be used, to ensure at least a certain amount of worker nodes are available at any time.

Reviewer: But IIUC it is not supported to have mixed Node Pools having on-demand and spot instances at the same time?

Author: ah true, forgot to straight that out

Author: fixed

Reviewer 2: But what's the configuration called when you set max price to `-1`? Doesn't that behave like a mixed one, even though it's not necessarily explicitly such?

Author: u can't set -1 :D

Reviewer 2: _Goes to read docs..._ :joy: 

Why?! [Azure docs](https://docs.microsoft.com/en-us/azure/virtual-machines/spot-vms#eviction-policy) specifically state this:
> If the max price is set to -1: The VM will not be evicted for pricing reasons. The max price will be the current price, up to the price for standard VMs. You will never be charged above the standard price.

For behavior that would be explicitly the desired outcome of "mixed" type node pool. I feel like having to put some arbitrary $highnumber is very confusing if not outright scary :man_shrugging:

Author: isn't it this setting: "Use current on-demand pricing as max"? :D i will make docs more specific about that

```"priority": "Spot",
                    "evictionPolicy": "Delete",
                    "billingProfile": {
                        "maxPrice": -1
                    }```

Author: Done

However, it is still possible that the node pool is in need of more worker nodes and there are just no spot instances available matching the request, based on the availability zone and the instance type(s). In this case the node pool cannot be scaled up. As a result, workloads will likely remain unscheduled.

To ensure enough capacity in a cluster, this guide presents a solution where

- two node pools exist: one configured to only use spot instances, the other one to only use on-demand instances.
- cluster-autoscaler is configured to scale up the spot node pool first, and only if that fails, scale up the on-demand node pool.

> Credit where credit is due: this solution has been proposed [in a blog post by Nir Forer](https://blog.doit-intl.com/running-eks-workloads-on-spot-instances-with-on-demand-instances-fallback-14bef39ce689). We adapt it here to match exactly the situation of Giant Swarm workload clusters on AWS.

Reviewer 3: ```suggestion
Credit where credit is due: this solution has been proposed [in a blog post by Nir Forer](https://blog.doit-intl.com/running-eks-workloads-on-spot-instances-with-on-demand-instances-fallback-14bef39ce689). We adapt it here to match exactly the situation of Giant Swarm workload clusters on Azure.
```

**Note:** This guide has been written for workload cluster release v14.1.0 with [cluster-autoscaler](https://github.com/kubernetes/autoscaler) v1.19.1, provided via our [cluster-autoscaler-app](https://github.com/giantswarm/cluster-autoscaler-app) version [v1.19.1](https://github.com/giantswarm/cluster-autoscaler-app/releases/tag/v1.19.1).

## Cluster set-up

First, create a cluster with at least version 14.1.x for the purpose of this tutorial.

Next, create two node pools.

1. Name: Spot,
   Enable spot instances: yes,
   Spot maximum price per hour: set your value for max price or use current on-demand pricing as max,
   Scaling range: min=1

2. Name: On-Demand,
   Instance type: the same as selected for the "Spot" node pool,
   Enable spot instances: no,
   Scaling range: min=1

If it is important to you that the nodes in both node pools run in the same availability zone(s), you have to select the matching zones manually for each node pool.

Make sure there are no other node pools in the cluster.

> Finally, once you have create the cluster and the two node pools, take note of their IDs. You will need them in the next steps.

Reviewer 2: ```suggestion
Finally, once you have created the cluster and the two node pools, take note of their IDs. You will need them in the next steps.
```

## Autoscaler expander configuration

First we create a ConfigMap manifest to provide configuration for the cluster-autoscaler's `priority` expander. This expander decides which node pool to scale up, based on a defined priority value.

Let's assume that our Spot node pool has the ID `sp0ti` and the on-demand node pool has ID `0ndmd`. We apply these two node pool ID in the ConfigMap as shown below:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-autoscaler-priority-expander
  namespace: kube-system
data:
  priorities: |-
    10:
      - .*0ndmd.*
    20:
      - .*sp0ti.*
```

> The `.data.priorities` part defines a map of priority values and regular expressions to select the Auto Scaling Group (ASG) associated with the node pool. Since with Giant Swarm the node pool ID is part of the respective Virtual Machine Scale Set (VMSS) name, we have to apply the IDs here.

Reviewer 2: ```suggestion
The `.data.priorities` part defines a map of priority values and regular expressions to select the Virtual Machine Scale Set (VMSS) associated with the node pool. Since with Giant Swarm the node pool ID is part of the respective VMSS name, we have to apply the IDs here.
```

  name: cluster-autoscaler-user-values
  namespace: abc12
data:
  values: |
    configmap:
      expander: priority
```

Save this manifest to e. g. `cluster-autoscaler-user-values.yaml` and apply it using

```nohighlight
kubectl apply -f ./cluster-autoscaler-user-values.yaml
```

## Test and observe

Back to the workload cluster, let's deploy some workloads. Here is a simple deployment manifest we can use for the purpose.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: nginx
  name: nginx
  namespace: default
spec:
  replicas: 10
  revisionHistoryLimit: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        imagePullPolicy: Always
        name: nginx
        resources:
          requests:
            cpu: "1"
            memory: 1Gi
      restartPolicy: Always
```

This would try to create 10 pods, where each one would request 1 CPU and 1 GiB of memory.

Once you apply this to the workload cluster, watch the autoscaler logs using this command:

```nohighlight
kubectl -n kube-system logs -l app=cluster-autoscaler -f
```

> The logs should show that cluster-autoscaler is registering the two Auto Scaling Groups (ASGs) belonging to our node pools, similar to this:

Reviewer 2: ```suggestion
The logs should show that cluster-autoscaler is registering the two Virtual Machine Scale Sets (VMSSs) belonging to our node pools, similar to this:
```

```nohighlight
I0701 11:10:16.028372       1 auto_scaling_groups.go:138] Registering ASG "nodepool-0ndmd"

> I0701 11:10:16.028394       1 auto_scaling_groups.go:138] Registering ASG "nodepool-sp0ti"

Reviewer 2: Are these possibly AWS specific?
